src/VirtualFile/VirtualPath.cs

```diff
@@ -11,7 +11,12 @@ namespace System.Web;
 
 internal sealed class VirtualPath
 {
-    private static IFileProvider Files => HttpRuntimeHelper.Services.GetRequiredService<IHostEnvironment>().ContentRootFileProvider;
+    private IFileProvider _fileProvider;
+    internal IFileProvider Files
+    {
+        get => _fileProvider ??= HttpRuntimeHelper.Services.GetRequiredService<IHostEnvironment>().ContentRootFileProvider;
```

Comment:
can you revert this? I don't want to try caching it outside of the httpRuntimeHelper. Long term I hope we have a better way to handle that, but for now, I'd rather it be retrieved this way

Reply:
The problem is BaseParser, BatchParser and TemplateParser all create a new VirtualPath at multiple places using the static VirtualPath.Create method. This requires us to update VirtualPath.Files with WebFormsFileProvider each time a new VirtualPath is created. To handle this, I had to make Files non-static and pass WebFormsFileProvider in an overloaded Create method.

Reply:
This is why I'd prefer within the parsers to provide the files to the Open/etc methods. That way, VirtualPath itself doesn't need to know where to get the file provider, the methods that need it get it passed to. This way, the caller can decide which file provider to use.